src/liballoc/collections/btree/set.rs

```diff
@@ -1055,6 +1086,46 @@ impl<'a, T> IntoIterator for &'a BTreeSet<T> {
     }
 }
 
+/// An iterator produced by calling `drain_filter` on BTreeSet.
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+pub struct DrainFilter<'a, T>(
+    btree_map::DrainFilter<'a, T, (), Box<dyn 'a + FnMut(&T, &mut ()) -> bool>>,
+)
+where
+    T: 'a + Ord;
+
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+impl<'a, T> Drop for DrainFilter<'a, T>
+where
+    T: 'a + Ord,
+{
+    fn drop(&mut self) {
+        self.for_each(drop);
```

Comment:
As written, I think this `Drop` is not necessary, because the inner `map::DrainFilter` will already do this. However, I'm going to suggest iterating more directly, and then this will be needed after all. Then consider #67290 as well.

Reply:
The way it's now, `DrainFilterInner` cannot iterate on its own because it doesn't have the predicate, so I think both `BTreeMap` and `BTreeSet` need to implement `Drop`. I don't understand what you mean with "iterating more directly".